lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -375,43 +418,162 @@ void UsdUndoAssignNewMaterialCommand::redo()
         auto addMaterialCmd
             = std::dynamic_pointer_cast<MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand>(*cmdsIt++);
         auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
-        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim());
+        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim(), _nodeId);
     }
 }
 
-void UsdUndoAssignNewMaterialCommand::connectShaderToMaterial(
-    Ufe::SceneItem::Ptr shaderItem,
-    UsdPrim             materialPrim)
+void UsdUndoAssignNewMaterialCommand::markAsFailed()
 {
-    auto                 shaderUsdItem = std::dynamic_pointer_cast<UsdSceneItem>(shaderItem);
-    auto                 shaderPrim = UsdShadeShader(shaderUsdItem->prim());
-    UsdShadeOutput       materialOutput;
-    PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+    _cmds->undo();
+    _cmds.reset();
+}
+
+UsdUndoAddNewMaterialCommand::UsdUndoAddNewMaterialCommand(
+    const UsdSceneItem::Ptr& parentItem,
+    const std::string&       nodeId)
+    : Ufe::InsertChildCommand()
+    , _parentPath((parentItem && parentItem->prim().IsActive()) ? parentItem->path() : Ufe::Path())
+    , _nodeId(nodeId)
+    , _cmds(std::make_shared<Ufe::CompositeUndoableCommand>())
+{
+}
+
+UsdUndoAddNewMaterialCommand::~UsdUndoAddNewMaterialCommand() { }
+
+UsdUndoAddNewMaterialCommand::Ptr
+UsdUndoAddNewMaterialCommand::create(const UsdSceneItem::Ptr& parentItem, const std::string& nodeId)
+{
+    // Changing the hierarchy of invalid items is not allowed.
+    if (!parentItem || !parentItem->prim().IsActive())
+        return nullptr;
+
```

Comment:
The other thing I'd like discussed in UFE: checking condition at creation time instead of execution time. The state can change between creation and execution, so verifying early can be detrimental because

- The condition can pass at creation but be invalid at execution, leading to an invalid state.
- The condition can fail at creation but would be valid at execution (happens in composite commands), which leads to denial of service when it should have worked.